### ISSUE TITLE
Service Worker offline on server

### ISSUE THREAD

**Issue by adnathanail** [angular/angular#25340]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
When running a build of my app locally with http-server the service worker functions exactly as expected, when going offline the site still loads and even after the server has been stopped the page still loads.
When running the exact same build on a server (nginx if relevant) the service worker seems to stop when the last tab with the site open on is closed. So if I have the site open in one tab and I open it again in another, and then make the second tab offline and refresh the app is cached. In fact I can even open it in 2 tabs, go offline in 1 and refresh, then go offline in the other and refresh and it works. It seems that even whilst offline the other tab is keeping the service worker alive however as soon as there is no tab using it (which happens briefly whilst reloading) the service worker self destructs! However, when only one tab with the site in is open and then turned offline and refreshed no requests seem to go to the service worker, the requests to the server fail and I get a 504.

## Expected behavior
Even when it is the only tab with the site open refreshing the page when offline should display the cached files from the service worker

## Minimal reproduction of the problem with instructions
Can't provide minimal reproduction because works when running locally

## What is the motivation / use case for changing the behavior?
Makes angular service worker function properly on server/whatever is causing the issue here

## Environment
<pre><code>
Angular version: 6.1.1

Browser:
- [x] Chrome (desktop) version 67.0.3396.99
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.11.1
- Platform:  Mac

Others:
I have a base-href set (to /d-a/) which I think is fine, I have put this in the required places in manifest.json and ngsw-config.json. Again this works locally.

Nginx config:
location /d-a/ {
        alias /home/redacted/redacted/d-a/;
        try_files $uri $uri/ $uri$args $uri$args/ /d-a/index.html;
}
When /d-a/ is requested try to find the requested file in /home/redacted/redacted/d-a/ if it's not there just serve /d-a/index.html (to let angular handle the routing)

Local server is http, remote is https.

Using @angular/material saw on some other issues about service-worker & pwa that this was potentially causing issues?
</code></pre>

**Comment by adnathanail** [comment]
I have literally changed nothing but now the local server does not work offline anymore

**Comment by AndreyLapchik** [comment]
I have a similar issue. 
Only on remote server, i get "Error when fetching the script" and "504 Bad Gateway". Happens principly with assets files, or files linked in css via url().
I tried making a brand new project, and going back to "@angular/cli": "~6.0.8", and "@angular/pwa": "^0.6.8" but it didn't help.

**Comment by gkalpak** [comment]
Sorry, but there is nothing we can do without a minimal reproduction. Happy to investigate when you can provide one.
(BTW, you might find some useful SW debugging info at `<your-domain>/ngsw/state`.)

**Comment by adnathanail** [comment]
Saw that suggested elsewhere but when I try i get:
<pre><code>
ERROR Error: Uncaught (in promise): Error: Cannot match any routes. URL Segment: 'ngsw/state'
Error: Cannot match any routes. URL Segment: 'ngsw/state'
</code></pre>

**Comment by gkalpak** [comment]
That probably means that the SW is not installed, in which case it can't indeed provide any debug info.

**Comment by AndreyLapchik** [comment]
Repro steps:
```
npm install --save @angular-cli@6.0.8
ng new app-test && cd app-test
ng add @angular/pwa@0.6.8 --project app-test
```

Put any image in assets.

In styles.css add : 
```
html {
    background-image: url(./assets/image.jpg);
}
```
Then
`ng build --prod`

Copy the content of the dist folder in your distant server. (Apache 2.4.9 in my case)
Go to the site.
Go offline.
Reload.
Background disapears and 504 Gateway Timeout for the image.

**Comment by adnathanail** [comment]
@gkalpak Tried with the versions you suggested [here](https://github.com/angular/angular-cli/issues/11640#issuecomment-410638864) as well as the ones in the comment above. Again works on local http-server but when on nginx (on https) offline does not function. I have now got ngsw/state to work and I get the following:
<pre><code>
[3m8s4u] TypeError(Failed to fetch, TypeError: Failed to fetch) Driver.fetch(https://da.redacted.co.uk/)
[2m59s194u] Error(Hash mismatch (cacheBustedFetchFromNetwork): https://da.redacted.co.uk/index.html: expected 32ca9784fa783ed51fd1f0e842b281fd951af65f, got fe7e41417229e8b86bdce275e4448d45f388830d (after cache busting), Error: Hash mismatch (cacheBustedFetchFromNetwork): https://da.redacted.co.uk/index.html: expected 32ca9784fa783ed51fd1f0e842b281fd951af65f, got fe7e41417229e8b86bdce275e4448d45f388830d (after cache busting)
    at PrefetchAssetGroup.<anonymous> (https://da.redacted.co.uk/ngsw-worker.js:674:35)
    at Generator.next (<anonymous>)
    at fulfilled (https://da.redacted.co.uk/ngsw-worker.js:304:62)) initializeFully for eb9fbe047e132f71917baab3deacdadd3879a015
[1m3s237u] TypeError(Failed to fetch, TypeError: Failed to fetch) Driver.fetch(https://da.redacted.co.uk/)
[58s203u] TypeError(Failed to fetch, TypeError: Failed to fetch) Driver.fetch(https://da.redacted.co.uk/ngsw.json?ngsw-cache-bust=0.6084465673480977)
[58s203u] Check for update aborted. (Client or server offline.) 
[1s814u] TypeError(Failed to fetch, TypeError: Failed to fetch) Driver.fetch(https://da.redacted.co.uk/)
</code></pre>
I believe I have fixed the first error about hashing, I have included it just in case it is relevant

**Comment by gkalpak** [comment]
@adnathanail, If you have fixed the hash mismatch error, then I don't see anything that doesn't work as expected (as far as I can tell by the log you posted).

##
@AndreyLapchik, yours is a different issue. It works if you use `url(/assets/image.jpg)` (using an absolute URL instead of a relative one), but that might break in other cases.

The problem is with how `@angular/cli` handles relative URLs. I.e. if you use `./assets/image.jpg` (relative URL), cli will create a top-level `image.<hash>.jpg` and use that in `url(...)` (instead of the one in assets). Since the SW does not cache that new image file, it is not available when offline.

You can of course update your `ngsw-config.json` to cache top-level images too, but I would argue this cli behavior breaks `@angular/service-worker`. Feel free to open an issue on the [@angular/cli][1] repo.

[1]: https://github.com/angular/angular-cli

**Comment by adnathanail** [comment]
@gkalpak so the "Check for updated aborted" and "Failed to fetch" are part of the normal running of the service worker? Because on the local server which is working fine doesn't have anything in ngsw/state, and those errors appear everytime I go offline

**Comment by gkalpak** [comment]
If you are offline (which seems to be the case based on the log), then it is expected that fetching files fails and checking for updates is aborted.

**Comment by adnathanail** [comment]
OK but then the service worker then fails to do its job and serve the cached files. Instead I get a 504

**Comment by gkalpak** [comment]
Again, I can't tell why without a minimal reproduction :grin:

**Comment by adnathanail** [comment]
@AndreyLapchik gave one earlier, obviously it's not that minimal because you have to put it on a real server but that's where the issue is...

**Comment by adnathanail** [comment]
Ahh sorry OK. This is quite long but:
1. `npm i -g @angular/cli@6.2.0-beta.1`
2. `ng new project`
3. `cd project`
4. `ng add @angular/pwa@0.8.0-beta.1`
5. `ng build --prod`
6. `cd dist/project/`
7. `http-server -p 8080`
8. Open `http://127.0.0.1:8080/`
9. Go offline (In dev tools or by killing the http-server) and refresh. It works fine
10. Upload the build to a server and serve it over https (preferably with nginx like this):
<pre><code>server {
    server_name project.redacted.co.uk www.project.redacted.co.uk;
    root /home/project/;
    location / {
        try_files $uri $uri$args $uri$args/ $uri/ /index.html =404;
    }
    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/project-0001/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/project-0001/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot
}</pre></code>
11. Go to the url of the server.
12. Go offline and refresh. HTTP 504
13. Open the same url in a new tab
14. Make either tab offline and refresh and it will serve the cached files, implying that the other open tab is maybe keeping the service worker alive or something? I don't really know how the service worker works..

**Comment by gkalpak** [comment]
I tried something similar and it seems to work. The fact that it works for you too using one HTTP server but breaks with another is an indication that something is wrong with your nginx configuration/setup.

Do you have a publicly accessible deployment on nginx that reproduces the error?

**Comment by adnathanail** [comment]
Do you mean like a link to it?

**Comment by gkalpak** [comment]
Yes.

**Comment by adnathanail** [comment]
Do you mind if I email you?

**Comment by gkalpak** [comment]
Nope.

**Comment by gkalpak** [comment]
For future reference:
@adnathanail and I were able to figure this out "offline".
It was an FTP client replacing `\n` newlines with `\r\n`, which caused the content-based hashes of the served files to not match with the expected hashes (that were computed with the old `\n` newlines).

**Comment by primg** [comment]
@gkalpak @adnathanail I have a similar issue, could you briefly explain how you found the issue, or better, how you fixed it ?

**Comment by adnathanail** [comment]
Sure the fundamental issue was that my FTP client was replacing newline characters. The problem with that is that angular CLI generates a hash of the generated code, which is check somewhere in the @angular/pwa code (No idea why..). If you go to your angular site and go to /ngsw/state, if this is your issue, you will see something like "hash mismatch" etc. No idea how gkalpak managed to figure out exactly where the mismatch was but essentially try and look around in all of your tools to see if anything is editing your code after you compile with angular/cli.

**Comment by primg** [comment]
Thank you very much. In the end my issue was somewhere else, but it's good to have it explained. Cheers 👍

**Comment by vytautas-pranskunas-** [comment]
I got has mismatch problem but for me it is intresting i got this error:
Error(Hash mismatch (cacheBustedFetchFromNetwork): https://localhost:4000/index.html: expected 8edd8e7e4544e9ff46eaf66b17072349b0a9bf33, got f04045b1e6bfc3cc652e92824812c63acbf5178c (after cache busting), Error: Hash mismatch (cacheBustedFetchFromNetwork): https://localhost:4000/index.html: expected 8edd8e7e4544e9ff46eaf66b17072349b0a9bf33, got f04045b1e6bfc3cc652e92824812c63acbf5178c (after cache busting)
    at PrefetchAssetGroup.<anonymous> (https://localhost:4000/ngsw-worker.js:675:35)
    at Generator.next (<anonymous>)
    at fulfilled (https://localhost:4000/ngsw-worker.js:304:62)) Error occurred while updating to manifest 719263827406523f2a7ec2f39e34fc27f9e0bcf9

in ngsw.json there is hash totally different 8edd8e7e4544e9ff46eaf66b17072349b0a9bf33

Any ideas how to solve it?

**Comment by vytautas-pranskunas-** [comment]
I was able to fix it
Originally service worker was registered by
`ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production }),`
but eventually it stopped registering it self so i added workaround registration with
`if (environment.production && 'serviceWorker' in navigator) {
            navigator.serviceWorker.getRegistration()
                .then(active => !active && navigator.serviceWorker.register('/ngsw-worker.js'))
                .catch(console.error);
        }`

this seems however broke or incorrectly registers service worker but it seems somehow registers it incorrectly.
So I figured out why SW is not registered with first line and it was because setInterval on my home page (https://github.com/angular/angular/issues/20970). I wrapped it to 
`
this.ngZone.runOutsideAngular(() => {
            this.slideIntervalId = setInterval(() => {
                this.ngZone.run(() => {
                   actual code here
                    this.cd.markForCheck();
               });
            }, 5000);
        });
`
and SW now is registered using first line. After checking /ngsw/state i see no errors all api requests are cached. 
Hope it helps somebody.

p.p.s. Wondering - does others like React or Vue has so much deeply hidden problems...

**Comment by abdatta** [comment]
@adnathanail  @gkalpak  Oh thank you so much. I had been spending my last 7 hours unable to understand why my PWA was not working offline! It was a simple "replacing \n newlines with \r\n" issue. I would have never found it without you guys! 😍  God bless you 👼

**Comment by GuilhermeLessa** [comment]
I dont understand. I need to configure the manifest manually for old browsers and safari? Or my Angular PWA project generated in CLI already does everything for me?

**Comment by GuilhermeLessa** [comment]
Thank's for your time, this questions is really complexed.

**Comment by gkalpak** [comment]
@GuilhermeLessa, your question doesn't seem to fall into the bug report or feature request category (which this issue tracker is suitable for). For support requests, please repost your issue on [StackOverflow](http://stackoverflow.com/) using tag `angular` or ask on [gitter](https://gitter.im/angular/angular). See also [CONTRIBUTING.md](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-got-a-question-or-problem).

If you are wondering why we don't resolve support issues via the issue tracker, please [check out this explanation](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-got-a-question-or-problem).

**Comment by Squidy06** [comment]
If anyone is experiencing problems with accessing _<your-domain>/ngsw/state_ (especially when using angular i18n):

From my observations, this can be accessed only from the root of the origin, as in the ngsw-worker.js you can see:

```
if (requestUrlObj.path === '/ngsw/state') {
    // Allow the debugger to handle the request, but don't affect SW state in any other way.
   event.respondWith(this.debugger.handleFetch(req));
   return;
}
```

So if you want it to work for example for <yourdomain>/fr/ then replace `'/ngsw/state'` with `'/fr/ngsw/state'`.

**Comment by gkalpak** [comment]
Unfortunately, it is currently not possible to get to `/ngsw/state` when the SW scope is not `/`.
This is a bug and will be fixed by #31082.

**Comment by Squidy06** [comment]
Strange, I couldn't access it, but after changing the thing I mentioned it started working.
